### PR TITLE
MULE-10253: support for custom types in smart connectors

### PR DESCRIPTION
**Pull request by fernandezlautaro**
Current smart connectors will have the possibility to define custom types for any input and output parameter.

### CODE REVIEW

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/loader/XmlExtensionLoaderDelegate.java [hunk 1]
```diff
@@ -71,14 +72,23 @@
   protected static final String CONFIG_NAME = "config";
 
   private static final ClassTypeLoader typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader();
-  private static final Map<String, MetadataType> types = ImmutableMap.<String, MetadataType>builder()
+  private static final Map<String, MetadataType> defaultInputTypes = ImmutableMap.<String, MetadataType>builder()
```
Comment (pablolagreca) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/loader/XmlExtensionLoaderDelegate.java]: Create a variable with the common set of types and the reuse it for defaultInputTypes and defaultOutputTypes

File: modules/extensions-xml-support/src/test/java/org/mule/runtime/extension/internal/loader/catalog/TypesCatalogLoaderTest.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import org.junit.Test;
+import org.mule.runtime.extension.internal.loader.catalog.loader.xml.TypesCatalogXmlLoader;
+import org.mule.runtime.extension.internal.loader.catalog.model.TypesCatalog;
+
+import java.io.File;
+
+/**
+ * TODO(fernandezlautaro: MULE-11501 this class must be moved to a separate module
+ */
+public class TypesCatalogLoaderTest {
```
Comment (pablolagreca) [inline modules/extensions-xml-support/src/test/java/org/mule/runtime/extension/internal/loader/catalog/TypesCatalogLoaderTest.java]: Must end with TestCase and extends AbstractMuleTestCase

File: modules/extensions-xml-support/src/test/resources/module-custom-types/module-custom-types-catalog.xml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (pablolagreca) [inline modules/extensions-xml-support/src/test/resources/module-custom-types/module-custom-types-catalog.xml]: This file has the same content as module-param-custom-types-catalog.xml. Can you use just one of them?

File: modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/loader/XmlExtensionLoaderDelegate.java (hunk 1 above)
Comment (fernandezlautaro) [inline modules/extensions-xml-support/src/main/java/org/mule/runtime/extension/internal/loader/XmlExtensionLoaderDelegate.java]: done

File: modules/extensions-xml-support/src/test/java/org/mule/runtime/extension/internal/loader/catalog/TypesCatalogLoaderTest.java (hunk 2 above)
Comment (fernandezlautaro) [inline modules/extensions-xml-support/src/test/java/org/mule/runtime/extension/internal/loader/catalog/TypesCatalogLoaderTest.java]: done

File: modules/extensions-xml-support/src/test/resources/module-custom-types/module-custom-types-catalog.xml (hunk 3 above)
Comment (fernandezlautaro) [inline modules/extensions-xml-support/src/test/resources/module-custom-types/module-custom-types-catalog.xml]: I'll change the one used in module-param-custom-types-catalog.xml, as using the same file will too much of a trouble for the parameterized test class.

**Pull request merged**